### PR TITLE
T22168: Support EknServices2 and EknServices3

### PR DESCRIPTION
**Pull request by smspillaz**
https://phabricator.endlessm.com/T22168

### CODE REVIEW

File: src/eos-companion-app-integration-helper.c [hunk 1]
```diff
@@ -470,6 +471,7 @@ lookup_eknservices_version (const gchar        *runtime_version,
     const gchar *eknservices_name;
     const gchar *search_provider_name;
   } runtime_to_eknservices_versions[] = {
+    { "1", "EknServices", "SearchProviderV1" },
```
Comment (erikos) [inline src/eos-companion-app-integration-helper.c]: Is this really desired? Quote/Word of the day both are against SDK1, also Vroom and the Service is not able to show the content using EKNServices1, if I set the runtime in the vroom manifest to 2 it works.

Word and Quote of the day we should probably blacklist.

File: src/eos-companion-app-service-app-info.c [hunk 2]
```diff
@@ -0,0 +1,214 @@
+/* Copyright 2017 Endless Mobile, Inc.
```
Comment (erikos) [inline src/eos-companion-app-service-app-info.c]: 2018

File: src/eos-companion-app-service-app-info.h [hunk 3]
```diff
@@ -0,0 +1,38 @@
+/* Copyright 2017 Endless Mobile, Inc.
```
Comment (erikos) [inline src/eos-companion-app-service-app-info.h]: 2018

**Comment by smspillaz** [comment]
> Not sure we want to support SDK1 apps.

@cosimoc asked if we could support them :-)

File: src/eos-companion-app-integration-helper.c (hunk 1 above)
Comment (smspillaz) [reply src/eos-companion-app-integration-helper.c]: I'm on the fence about blacklisting those two. Technically we have a working content app for both of them, but they aren't really meant to be user facing.

Since both of them are actually set as `NoDisplay=True` on the desktop side, I wonder if we should use that as the criteria to remove them from `/list_applications`. Thoughts @cosimoc ?

Comment (smspillaz) [reply src/eos-companion-app-integration-helper.c]: > Vroom and the Service is not able to show the content using EKNServices1, if I set the runtime in the vroom manifest to 2 it works.

That is very curious. How did it work before if Vroom was an SDK1 app? It couldn't have been supported with EknServices2 ....

**Comment by smspillaz** [comment]
@cosimoc It appears that the SDK1 support was accidentally removed. I'll quickly re-test this but it should be good to go otherwise.

File: eoscompanion/applications_query.py [hunk 4]
```diff
@@ -60,15 +60,22 @@ def application_listing_from_app_info(app_info):
                               display_name,
                               short_description,
                               icon,
-                              language)
+                              language,
+                              eknservices_name,
+                              search_provider_name)
```
Comment (cosimoc) [inline eoscompanion/applications_query.py]: This hunk is unrelated to this commit

File: src/eos-companion-app-integration-helper.c [hunk 5]
```diff
@@ -506,10 +510,8 @@ list_application_infos (GCancellable  *cancellable,
           if (!is_compatible_app)
             continue;
 
-          if (!load_desktop_info_key_file_for_app_id (app_name, &app_info, error))
-            return NULL;
+          app_info = load_desktop_info_key_file_for_app_id (app_name, &check_app_error);
```
Comment (cosimoc) [inline src/eos-companion-app-integration-helper.c]: I don't believe `check_app_error` exists by the time it gets introduced with this commit.

File: src/eos-companion-app-integration-helper.c [hunk 6]
```diff
@@ -486,6 +486,7 @@ list_application_infos (GCancellable  *cancellable,
           g_autofree gchar *runtime_name = NULL;
```
Comment (cosimoc) [inline src/eos-companion-app-integration-helper.c]: This has to be changed to `runtime_spec` as part of this commit as well.

File: src/eos-companion-app-integration-helper.c [hunk 7]
```diff
@@ -608,26 +658,61 @@ eos_companion_app_service_finish_load_application_icon_data_async (GAsyncResult
...
+
 static void
 load_application_info_thread (GTask        *task,
                               gpointer      source,
                               gpointer      task_data,
                               GCancellable *cancellable)
 {
   g_autoptr(GError) local_error = NULL;
-  g_autoptr(GDesktopAppInfo) app_info = NULL;
+  g_autoptr(EosCompanionAppServiceAppInfo) info = load_application_info((const gchar *) task_data,
```
Comment (cosimoc) [inline src/eos-companion-app-integration-helper.c]: Missing space before paren

File: src/eos-companion-app-integration-helper.c [hunk 8]
```diff
@@ -608,26 +658,61 @@ eos_companion_app_service_finish_load_application_icon_data_async (GAsyncResult
...
+                       GError      **error)
+{
+  g_autoptr(GDesktopAppInfo) app_info = NULL;
+  g_autofree gchar *runtime_name = NULL;
+  g_autofree gchar *runtime_version = NULL;
+  const gchar *eknservices_name = NULL;
+  const gchar *search_provider_name = NULL;
+  g_autofree gchar *runtime_spec = NULL;
+
+  runtime_spec = eos_companion_app_service_get_runtime_name_for_app_id (app_id, error);
```
Comment (cosimoc) [inline src/eos-companion-app-integration-helper.c]: It's confusing that this function says that it returns a `runtime_name`, yet you call the variable `runtime_spec` and you need to call `parse_runtime_spec()` to actually get something called `runtime_name`

File: src/eos-companion-app-integration-helper.c [hunk 9]
```diff
@@ -608,26 +658,61 @@ eos_companion_app_service_finish_load_application_icon_data_async (GAsyncResult
...
+
+  if (runtime_spec == NULL)
+    return FALSE;
+
+  if (!parse_runtime_spec (runtime_spec, &runtime_name, &runtime_version, error))
```
Comment (cosimoc) [inline src/eos-companion-app-integration-helper.c]: It does not look like `runtime_name` is used, could you pass `NULL` instead?

File: src/eos-companion-app-integration-helper.c [hunk 10]
```diff
@@ -456,6 +457,41 @@ load_desktop_info_key_file_for_app_id (const gchar      *app_id,
...
+                     runtime_to_eknservices_versions[i].runtime_version) == 0)
+        {
+          *out_eknservices_name = runtime_to_eknservices_versions[i].eknservices_name;
+          *out_search_provider_name = runtime_to_eknservices_versions[i].search_provider_name;
+          return;
+        }
+    }
+
+  *out_eknservices_name = NULL;
+  *out_search_provider_name = NULL;
```
Comment (cosimoc) [inline src/eos-companion-app-integration-helper.c]: Three things about this function:
- it's not typical to see functions with out params that return void, you should probably return a boolean that indicates whether this succeeded
- it's not typical for functions to touch out params in case of failure (other than a gerror param)
- by not copying the strings in the out params, you're extending the scope of const strings outside of where they're declared. Even if it may work, I believe that's undefined behavior and I had to spend a ton of time debugging an issue last week that was caused just by that: https://git.gnome.org/browse/sushi/commit/?id=0e7856afb9f2194579cfc0e5545f2a402df6f020. Please copy those strings :-)

File: eoscompanion/routes.py [hunk 11]
```diff
@@ -1136,19 +1207,45 @@ def _on_got_metadata_callback(load_metadata_error, metadata_bytes):
...
+                                               callback=_on_got_shards_callback)
+
     logging.debug(
         'Get content metadata: clientId=%s, '
         'applicationId=%s, contentId=%s',
         query['deviceUUID'],
         query['applicationId'],
-        query['contentId'])
+        query['contentId']
+    )
```
Comment (cosimoc) [inline eoscompanion/routes.py]: pylint did not like the indentation you're proposing here, and I had to change it to this style in order to fix the build the other day. See https://github.com/endlessm/eos-companion-app-integration/commit/8e8071fdf80d62e295a9abb2ddbddb4b9741ee2e

Comment (smspillaz) [reply eoscompanion/routes.py]: Odd... I am pretty sure I had to do this to get pylint not to complain. I'll take another look.

File: src/eos-companion-app-integration-helper.c (hunk 6 above)
Comment (smspillaz) [reply src/eos-companion-app-integration-helper.c]: Agreed

File: src/eos-companion-app-integration-helper.c (hunk 10 above)
Comment (smspillaz) [reply src/eos-companion-app-integration-helper.c]: I'll make the function return a boolean.

As for the string lifetime, that *should* work because the strings have static lifetime. However, I can also just copy them at the cost of a negligible amount of memory usage

File: src/eos-companion-app-integration-helper.c (hunk 8 above)
Comment (smspillaz) [reply src/eos-companion-app-integration-helper.c]: Yeah, I'll rename it

File: src/eos-companion-app-integration-helper.c (hunk 7 above)
Comment (smspillaz) [reply src/eos-companion-app-integration-helper.c]: It strikes again!

File: src/eos-companion-app-service-app-info.c [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+
+static void
+eos_companion_app_service_app_info_finalize (GObject *object)
+{
+  EosCompanionAppServiceAppInfo *store = EOS_COMPANION_APP_SERVICE_APP_INFO (object);
+  EosCompanionAppServiceAppInfoPrivate *priv = eos_companion_app_service_app_info_get_instance_private (store);
+
+  g_clear_object (&priv->desktop_app_info);
+  g_clear_pointer (&priv->eknservices_name, g_free);
+  g_clear_pointer (&priv->search_provider_name, g_free);
```
Comment (smspillaz) [inline src/eos-companion-app-service-app-info.c]: Oh wow, autocode stikes again - we were freeing static memory here -__-"

Comment (smspillaz) [reply src/eos-companion-app-service-app-info.c]: Or in fact, we were copying it

File: src/eos-companion-app-integration-helper.c [hunk 13]
```diff
@@ -500,11 +501,17 @@ list_application_infos (GCancellable  *cancellable,
...
+              g_message ("Flatpak at %s has a damaged installation and checking "
+                         "its metadata failed with: %s, ignoring",
+                         flatpak_directory,
+                         check_app_error->message);
+              continue;
+            }
 
           /* Check if the application is an eligible content app */
-          if (!app_is_compatible (app_name, runtime_name, &is_compatible_app, error))
+          if (!app_is_compatible (app_name, runtime_spec, &is_compatible_app, error))
```
Comment (smspillaz) [inline src/eos-companion-app-integration-helper.c]: Actually I'll change it back and do that in a later commit

File: eoscompanion/applications_query.py (hunk 4 above)
Comment (smspillaz) [reply eoscompanion/applications_query.py]: This has now been fixed

File: src/eos-companion-app-integration-helper.c (hunk 5 above)
Comment (smspillaz) [reply src/eos-companion-app-integration-helper.c]: This should also have been fixed now

File: eoscompanion/routes.py (hunk 11 above)
Comment (smspillaz) [reply eoscompanion/routes.py]: Yeah, pylint doesn't complain with this style, at least not from within my own flatpak builds?

Comment (smspillaz) [reply eoscompanion/routes.py]: Ah, the pylint errors that you're referring to are because the later arguments had more indentation than the former ones (the end paren placement is actually flexible)

File: src/eos-companion-app-integration-helper.c (hunk 6 above)
Comment (cosimoc) [reply src/eos-companion-app-integration-helper.c]: It still hasn't been addressed. In the commit `runtime_spec` is used, but the variable does not exist.

File: src/eos-companion-app-integration-helper.c (hunk 7 above)
Comment (cosimoc) [reply src/eos-companion-app-integration-helper.c]: And hasn't been fixed yet! :)
